library/std/src/sync/mutex.rs

```diff
@@ -125,11 +125,16 @@ use crate::sys::locks as sys;
 ///     let res_mutex_clone = Arc::clone(&res_mutex);
 ///
 ///     threads.push(thread::spawn(move || {
-///         let mut data = data_mutex_clone.lock().unwrap();
-///         // This is the result of some important and long-ish work.
-///         let result = data.iter().fold(0, |acc, x| acc + x * 2);
-///         data.push(result);
-///         drop(data);
+///         // Here we use a block to limit the lifetime of the lock guard.
+///         let result = {
+///             let mut data = data_mutex_clone.lock().unwrap();
+///             // This is the result of some important and long-ish work.
+///             let result = data.iter().fold(0, |acc, x| acc + x * 2);
+///             data.push(result);
+///             result
+///             // the mutex guard gets dropped here, together with any other values
+///             // created in the critical section.
+///         };
 ///         *res_mutex_clone.lock().unwrap() += result;
```

Comment:
I think it's worth commenting that this `res` guard is a temporary, dropped at the end of this statement.